### ISSUE TITLE
Suggestion: add `--avoid-file-write` compiler option

### ISSUE THREAD

**Issue by noomorph** [Microsoft/TypeScript#16242]
**Desired TypeScript Version to land in:** 2.4.x / 2.5.x (?)

This suggestion comes from real-life issues of integrating TypeScript command line compiler with tools supporting `watch` mode.

Let's say I have `webpack --watch` running in one tab, and in the other tab I execute a single run: `tsc`.

Since we have this line: https://github.com/Microsoft/TypeScript/blob/master/src/compiler/program.ts#L166 — `tsc` is going to rewrite all output files. Webpack and many other watchers simply love file timestamps and their caching algorithms are usually quite dependent on the "modified time".

If all output files are rewritten, then it makes the watcher think that all files have been changed and here we lose a potential performance gain because watcher rereads them again. This was my case with `webpack/lib/CachePlugin`.

Also, on its single (first) run, `tsc` starts with an empty map of output fingerprints and it never tries to build them on first comparison here:

https://github.com/Microsoft/TypeScript/blob/master/src/compiler/program.ts#L141

Therefore, when we are running for the first time (even in watch mode, not to say of the regular compile mode), we always rewrite the output files even if they are the same. I'm not saying this is invalid, it's just a pity that there is no flag to customize the compiler's behavior in this case.

I was looking how to improve `webpack` with enabled caching on one of big private projects I'm working on (over 1000 typescript files which are compiled with `tsc` and then separately bundled `webpack`), so I forked the TypeScript project and added my `avoidFileWrite` flag there:

https://github.com/noomorph/TypeScript/commit/57eb3f45b3d0cfd791cf630a51f446412ea4e85a

My results in my case were:

```
MY FORK (--avoid-file-write)

ts:build                     12.4s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 10%
webpack(hooked4312):webpack  14.3s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 11%
ts:build                     12.7s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 10%
webpack(hooked4312):webpack   3.4s  ▇▇▇▇▇ 3%
ts:build                     12.5s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 10%
webpack(hooked4312):webpack   3.5s  ▇▇▇▇▇ 3%
ts:build                     12.7s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 10%
webpack(hooked4312):webpack   3.3s  ▇▇▇▇▇ 3%

TYPESCRIPT 2.3.4

ts:build                     12.6s  ▇▇▇▇▇▇▇▇▇▇▇▇▇ 8%
webpack(hooked4675):webpack  13.2s  ▇▇▇▇▇▇▇▇▇▇▇▇▇ 8%
ts:build                     12.6s  ▇▇▇▇▇▇▇▇▇▇▇▇▇ 8%
webpack(hooked4675):webpack   7.2s  ▇▇▇▇▇▇▇ 4%
ts:build                     12.7s  ▇▇▇▇▇▇▇▇▇▇▇▇▇ 8%
webpack(hooked4675):webpack   6.3s  ▇▇▇▇▇▇▇ 4%
ts:build                     12.7s  ▇▇▇▇▇▇▇▇▇▇▇▇▇ 8%
webpack(hooked4675):webpack   6.4s  ▇▇▇▇▇▇▇ 4%
ts:build                     12.8s  ▇▇▇▇▇▇▇▇▇▇▇▇▇ 8%
webpack(hooked4675):webpack   6.2s  ▇▇▇▇▇▇▇ 4%
```

So I was able to decrease the webpack watcher execution time from 6.5s to 3.5s on my configuration.

My fork is not ready for the merging but if you are positive about this suggestion — I can make a pull request out of it in some nearest future. Better naming ideas for the suggested compiler option are also gladly accepted.

**Comment by jwbay** [comment]
Any reason you have webpack reading the JS output files from tsc instead of the TS files themselves? Webpack loaders can compile files on the fly with no intermediate file on disk. http://www.typescriptlang.org/docs/handbook/integrating-with-build-tools.html#webpack

**Comment by ikatyang** [comment]
I think the point here may be

1. they have some files that will be bundled twice or more, so that they don't want to compile twice.
2. they want to have fully type-checking since I realized that things like loader would just compile file one by one instead of compile using fully types. (correct me if I'm wrong...)

**Comment by noomorph** [comment]
@jwbay , @ikatyang , thanks for participating in the discussion.

@jwbay , imagine a bigger picture — what if I substitute `webpack --watch` in this equation with `karma --watch` or any other tool in watch mode. Usually, every such tool with a watch mode (personally, I've checked `karma`, for example) suffers from the same kind of problem — they pay a big attention to timestamps, so this suggestion is relevant not only to `webpack` but to a broader range of tools in watch mode.

@ikatyang , indeed guessed things. Currently, there are seven bundles in the project that include shared files too. And you are right that I would prefer to have a full type checking since it does not take too long (namely, 12s in cold start mode, but I checked that potentially we can improve to 6s in `tsc --watch` or similar mode).

I'm not saying that it is not possible to achieve performance boost in every particular use case but it would be very nice to add "power user options" (like the one I suggest) for those people who are responsible for setting up build configurations in their projects because circumstances may vary from a project to a project.
